spec/instrumental_reporters_spec.rb

```diff
@@ -5,29 +5,212 @@
+        @agent.should_receive(:gauge).with("queue.process", anything)
+
+        # typically instrumental reporters would be loaded in an initalizer
+        # and this _extend_ could be done inside the job itself, but here
+        # we are in a weird situation.
+        TestResqueJob.send(:extend, Instrumental::ResquePlugin)
+        Resque.enqueue(TestResqueJob, { "success" => true })
+      end
+
+      specify "job errors are instrumented" do
+        @agent.stub(:increment)
+        @agent.should_receive(:increment).with("queue.error", 1)
+
+        # typically instrumental reporters would be loaded in an initalizer
+        # and this _extend_ could be done inside the job itself, but here
+        # we are in a weird situation.
+        TestResqueJob.send(:extend, Instrumental::ResquePlugin)
+        lambda { Resque.enqueue(TestResqueJob, { "error" => true }) }.should raise_error(StandardError)
+      end
+    end
+
+    describe "sidekiq" do
+      before do
+        require "sidekiq"
+        require 'sidekiq/testing'
+        Sidekiq::Testing.inline!
+        InstrumentalReporters.activate
+        # sidekiq doesn't use middleware by default in their testing
+        # harness, so we add it just as metrician does
+        Sidekiq::Testing.server_middleware do |chain|
+          chain.add Instrumental::SidekiqMiddleware
+        end
+        @agent = InstrumentalReporters.agent
+      end
+
+      specify "Sidekiq is instrumented" do
+        @agent.stub(:gauge)
+        @agent.should_receive(:gauge).with("queue.process", anything)
+
+        # avoid load order error of sidekiq here by just including the
+        # worker bits at latest possible time
+        TestSidekiqWorker.send(:include, Sidekiq::Worker)
+        TestSidekiqWorker.perform_async({ "success" => true})
+      end
+
+      specify "job errors are instrumented" do
+        @agent.stub(:increment)
+        @agent.should_receive(:increment).with("queue.error", 1)
+
+        # avoid load order error of sidekiq here by just including the
+        # worker bits at latest possible time
+        TestSidekiqWorker.send(:include, Sidekiq::Worker)
+        lambda { TestSidekiqWorker.perform_async({ "error" => true}) }.should raise_error(StandardError)
+      end
+    end
+  end
+
+  describe "cache systems" do
+    specify "redis is instrumented" do
+      require "redis"
+      InstrumentalReporters.activate
+
+      client = Redis.new
+      agent = InstrumentalReporters.agent
+      agent.stub(:gauge)
+      agent.should_receive(:gauge).with("cache.command", anything)
+      client.get("foo-#{rand(100_000)}")
+    end
+
+    # Why is this fuckshow? In Rails 4x, we're going to load the
+    # memcached gem, and in 5 we're going to use dalli
+    # this will allow us to set up a client without blowing up
+    # in the wrong env.
+    def memcached_client
+      begin
+        require "memcached"
+        return Memcached.new("localhost:11211")
+      rescue LoadError
```

Comment:
There's a case that my code covered that your does not, which is that we are in a gemset where neither gem exists. In the original it would have been vociferous, and the new code it will silently just not run any tests, I think.